process/block/metablock.go

```diff
@@ -1854,106 +1779,66 @@ func (mp *metaProcessor) MarshalizedDataToBroadcast(
 	return mrsData, mrsTxs, nil
 }
 
-func (mp *metaProcessor) getSortedHeadersPerShard(round uint64) (map[uint32][]*hashAndHdr, int, error) {
-	shardBlocksPool := mp.dataPool.ShardHeaders()
-	if shardBlocksPool == nil {
-		return nil, 0, process.ErrNilShardBlockPool
-	}
-
-	mp.mutNotarizedHdrs.RLock()
-	if mp.notarizedHdrs == nil {
-		mp.mutNotarizedHdrs.RUnlock()
-		return nil, 0, process.ErrNotarizedHdrsSliceIsNil
-	}
-
-	hashAndBlockMap := make(map[uint32][]*hashAndHdr)
-	for _, key := range shardBlocksPool.Keys() {
-		val, _ := shardBlocksPool.Peek(key)
-		if val == nil {
-			continue
-		}
-
-		hdr, ok := val.(*block.Header)
-		if !ok {
-			continue
-		}
-
-		if hdr.GetRound() > round {
-			continue
-		}
+func (mp *metaProcessor) getLongestShardsChainFromLastNotarized(round uint64) ([]data.HeaderHandler, [][]byte, map[uint32][]data.HeaderHandler, error) {
+	lastCrossNotarizedHeader := make(map[uint32]data.HeaderHandler, mp.shardCoordinator.NumberOfShards())
+	hdrsMap := make(map[uint32][]data.HeaderHandler, 0)
+	hdrsHashesMap := make(map[uint32][][]byte, 0)
 
-		currShardId := hdr.ShardId
-		if mp.lastNotarizedHdrForShard(currShardId) == nil {
-			continue
-		}
-
-		if hdr.GetRound() <= mp.lastNotarizedHdrForShard(currShardId).GetRound() {
-			continue
-		}
-
-		if hdr.GetNonce() <= mp.lastNotarizedHdrForShard(currShardId).GetNonce() {
-			continue
+	// save last committed header for verification
+	for shardID := uint32(0); shardID < mp.shardCoordinator.NumberOfShards(); shardID++ {
+		lastCrossNotarizedHeaderForShard, _, err := mp.blockTracker.GetLastCrossNotarizedHeader(shardID)
+		if err != nil {
+			return nil, nil, nil, err
 		}
 
-		hashAndBlockMap[currShardId] = append(hashAndBlockMap[currShardId],
-			&hashAndHdr{hdr: hdr, hash: key})
+		lastCrossNotarizedHeader[shardID] = lastCrossNotarizedHeaderForShard
 	}
-	mp.mutNotarizedHdrs.RUnlock()
 
-	maxHdrLen := mp.sortShardHeadrs(hashAndBlockMap)
-	return hashAndBlockMap, maxHdrLen, nil
-}
-
-func (mp *metaProcessor) sortShardHeadrs(hashAndBlockMap map[uint32][]*hashAndHdr) int {
 	maxHdrLen := 0
-	for shardId := uint32(0); shardId < mp.shardCoordinator.NumberOfShards(); shardId++ {
-		hdrsForShard := hashAndBlockMap[shardId]
-		if len(hdrsForShard) == 0 {
-			continue
-		}
+	for shardID := uint32(0); shardID < mp.shardCoordinator.NumberOfShards(); shardID++ {
+		hdrsForShard, hdrsHashesForShard := mp.blockTracker.ComputeLongestChain(shardID, lastCrossNotarizedHeader[shardID])
```

Comment:
I would make unit tests for the tracker component in next PR (cross mini blocks balancer)

Reply:
that is already a new implementation, which will need his own unit tests. It will be a big PR as well. I am curios about the performance of the computeLongestChain function when there is a gap of 1000 blocks between shard and meta, and there is a shard split.

Reply:
Actually this new PR would be very short and the biggest part of implementation would be done exactly in block tracker where also the tests would be done. The rest of the PR would be almost nothing, 2 lines (a continue) in CreateMbsAndProcessTransactionsFromMe if blockTracker.IsShardStuck(shardID) would return true. 

About the computation time for longest chain, I already had a case in the system tests, with 6000 height for longest chain because of one missing header and everything ran instantly on each call. This is why in the last commit  I have added a requestHandler to the blockTracker to can request for missing stuff. I would also add a time measurement in block processor for the call of ComputeLongestChain just to check be sure that the computation time would not exceeded some reasonable limits.